modules/gapi/src/streaming/onevpl/engine/preproc/preproc_engine.cpp

```diff
@@ -0,0 +1,501 @@
+        return pp_session::create<EngineSession>(nullptr);
+    }
+
+    // recalculate size param according to VPP alignment
+    mfxVPPParams.vpp.Out.Width = ALIGN16(mfxVPPParams.vpp.Out.Width);
+    mfxVPPParams.vpp.Out.Height = ALIGN16(mfxVPPParams.vpp.Out.Height);
+    mfxVPPParams.vpp.Out.CropW = mfxVPPParams.vpp.Out.Width;
+    mfxVPPParams.vpp.Out.CropH = mfxVPPParams.vpp.Out.Height;
+
+    GAPI_LOG_DEBUG(nullptr, "\nFrom:\n{\n" << mfx_frame_info_to_string(mfxVPPParams.vpp.In) <<
+                            "}\nTo:\n{\n" << mfx_frame_info_to_string(mfxVPPParams.vpp.Out) << "}");
+
+    // find existing session
+    GAPI_LOG_DEBUG(nullptr, "Find existing vpp_pp_session for requested frame params"
+                            ", total sessions: " << preproc_session_map.size());
+    auto it = preproc_session_map.find(mfxVPPParams.vpp.In);
+    if (it != preproc_session_map.end()) {
+        GAPI_LOG_DEBUG(nullptr, "[" << it->second->session << "] found");
+        return pp_session::create(std::static_pointer_cast<EngineSession>(it->second));
+    }
+
+    // NB: make some sanity checks
+    IDeviceSelector::DeviceScoreTable devices = acceleration_policy->get_device_selector()->select_devices();
+    GAPI_Assert(devices.size() == 1 && "Multiple(or zero) acceleration devices case is unsupported");
+    AccelType accel_type = devices.begin()->second.get_type();
+    // assign acceleration
+    if (accel_type == AccelType::DX11) {
+        mfxVPPParams.IOPattern = MFX_IOPATTERN_IN_VIDEO_MEMORY | MFX_IOPATTERN_OUT_VIDEO_MEMORY;
+    } else {
+        mfxVPPParams.IOPattern = MFX_IOPATTERN_IN_SYSTEM_MEMORY | MFX_IOPATTERN_OUT_SYSTEM_MEMORY;
+    }
+
+    // clone existing VPL session to inherit VPL loader configuration
+    // and avoid refer to any global state
+    // TODO no clone due to clone issue
+
+    mfxSession mfx_vpp_session = params.handle;
+    mfxStatus sts = MFX_ERR_NONE;
+
+    // TODO: simply use clone after VPL bug fixing
+    //sts = MFXCloneSession(params.handle, &mfx_vpp_session);
+    sts = MFXCreateSession(mfx_handle, impl_number, &mfx_vpp_session);
+    if (sts != MFX_ERR_NONE) {
+        GAPI_LOG_WARNING(nullptr, "Cannot clone VPP session, error: " << mfxstatus_to_string(sts));
+        GAPI_Assert(false && "Cannot continue VPP preprocessing");
+    }
+
+    sts = MFXJoinSession(params.handle, mfx_vpp_session);
+    if (sts != MFX_ERR_NONE) {
+        GAPI_LOG_WARNING(nullptr, "Cannot join VPP sessions, error: " << mfxstatus_to_string(sts));
+        GAPI_Assert(false && "Cannot continue VPP preprocessing");
+    }
+
+    GAPI_LOG_INFO(nullptr, "[" << mfx_vpp_session << "] starting pool allocation");
+    VPLAccelerationPolicy::pool_key_t vpp_out_pool_key {};
+    try {
+        // assing HW acceleration processor
+        acceleration_policy->init(mfx_vpp_session);
+        try {
+            // ask to allocate external memory pool
+            mfxFrameAllocRequest vppRequests[2];
```

Comment:
I have lost thread. Why is 2?

Reply:
it is MFX/VPL API: `MFXVideoVPP_QueryIOSurf` returns array of 2 elements

it is possible to add naming variable...